### PR TITLE
New tools/certs Makefile targets for vm, istioctl, and token

### PR DESCRIPTION
**Pull request by esnible**
No work item.

- This provides the new target `${NAMESPACE}-certs-vm` for VM certificates
- This provides the new target `${NICKNAME}-certs-istioctl for `istioctl` certificates
- This provides the new target `%/token-cert` for tokens.  This work is incomplete.

To test the istioctl certificates, do

```
ISTIO_DIR=~/go/src/istio.io/istio
NICKNAME=mycluster
cd /tmp
make --makefile $ISTIO_DIR/tools/certs/Makefile $NICKNAME-certs-istioctl
istioctl x version --xds-address localhost:15012 --authority istiod.istio-system.svc --cert-dir $NICKNAME-istioctl
```

cc @costinm @howardjohn @sdake

### CODE REVIEW

**Comment by esnible** [comment]
The new **targets** in this _Makefile_ are used by [Istioctl troubleshooting without control plane access for Central Istiod](https://github.com/istio/istio/wiki/Istioctl-troubleshooting-without-control-plane-access-for-Central-Istiod) which I have placed on the Wiki.  Anyone really wishing to try out the new targets can use that tutorial.

File: tools/certs/Makefile [hunk 1]
```diff
@@ -277,3 +279,93 @@ root-key.pem:
...
+# CA_ADDR can be set to a different location - reflecting the gateway address
+CA_ADDR ?= istiod.istio-system.svc:15012
+# Get a token
+%/token:
+	# Using local .kube/config credentials to get a JWT token.
+	echo '{"kind":"TokenRequest","apiVersion":"authentication.k8s.io/v1","spec":{"audiences":["istio-ca"], "expirationSeconds":2592000}}' | \
+			kubectl create --raw /api/v1/namespaces/$(dir $@)serviceaccounts/default/token -f - | \
+			jq -j '.status.token' > $(dir $@)istio-token
+	# Get the Istid root CA (only for self-signed)
+	kubectl -n istio-system get secret istio-ca-secret -ojsonpath='{.data.ca-cert\.pem}' | \
```
Comment (linsun) [inline tools/certs/Makefile:338]: I think we are assuming the secret exists in istio-system ns which is not always the case.

File: tools/certs/Makefile [hunk 2]
```diff
@@ -277,3 +279,93 @@ root-key.pem:
...
+	@echo "O = $(INTERMEDIATE_ORG)" >> $@
+	@echo "CN = $(WORKLOAD_CN)" >> $@
+	@echo "L = $(L:/=)" >> $@
+
+#------------------------------------------------------------------------
+
+## <namespace>/token-cert will get a certificate for namespace using the token method. This only requires namespace
+# permissions, will not access istio-system or the certifcate.
+# CA_ADDR can be set to a different location - reflecting the gateway address
+CA_ADDR ?= istiod.istio-system.svc:15012
```
Comment (linsun) [inline tools/certs/Makefile:330]: is CA_ADDR the key user config that I can leverage in the case I have istiod1 (mesh1) in ns 1, and istiod2 (mesh2) in ns2, both are on c1?

File: tools/certs/Makefile (hunk 1 above)
Comment (linsun) [reply tools/certs/Makefile:338]: Also, in central istiod env, the istio-ca-secret doesn't exist in istiod cluster.  It exists in the workload cluster.

Comment (linsun) [reply tools/certs/Makefile:338]: Think the error below related to it.  
```
$ ISTIO_NAMESPACE=istio2 make --makefile ../istio/tools/certs/Makefile linistio10-certs-istioctl  
fetching root ca from k8s cluster: mycluster-wdc06-b3c.4x16-bs76p72w0qc3sa93oti0
Error from server (NotFound): secrets "istio-ca-secret" not found
Error from server (NotFound): secrets "istio-ca-secret" not found
generating linistio10/ca-key.pem
Generating RSA private key, 4096 bit long modulus
...................................++
......................................................................++
e is 65537 (0x10001)
generating linistio10/cluster-ca.csr
generating linistio10/ca-cert.pem
Signature ok
subject=/O=Istio/CN=Intermediate CA/L=linistio10
unable to load certificate
4591312364:error:09FFF06C:PEM routines:CRYPTO_internal:no start line:/BuildRoot/Library/Caches/com.apple.xbs/Sources/libressl/libressl-47.11.1/libressl-2.8/crypto/pem/pem_lib.c:684:Expecting: TRUSTED CERTIFICATE
make[1]: *** [linistio10/ca-cert.pem] Error 1
rm linistio10/cluster-ca.csr linistio10/intermediate.conf
make: *** [linistio10-certs-istioctl] Error 2
(⎈ |mycluster-wdc06-b3c.4x16/bs76p72w0qc3sa93oti0:default)
~/go/src/istio.io/testingdir ⌚ 21:51:08
$ k get secret -n istio2       
NAME                                                             TYPE                                  DATA   AGE
default-token-7dqrs                                              kubernetes.io/service-account-token   3      29d
istio-kubeconfig                                                 Opaque                                1      14d
istiod-service-account-token-td46l                               kubernetes.io/service-account-token   3      29d
mycluster-wdc06-b-406454-85f044fc29ce613c264409c04a76c95d-0001   Opaque                                2      28d
(⎈ |mycluster-wdc06-b3c.4x16/bs76p72w0qc3sa93oti0:default)
~/go/src/istio.io/testingdir ⌚ 21:51:29
$ k get pods -n istio2      
NAME                      READY   STATUS    RESTARTS   AGE
istiod-7b6565d7d5-jcbhb   1/1     Running   0          14d
```

Comment (linsun) [reply tools/certs/Makefile:338]: if i run this on the workload cluster, it doesn't work either if I use a special ns.  (I think it worked earlier when I didn't specify ns)

```
⎈ |linistio10/7caab3af9f514f028081a8180c107b69:default)
~/go/src/istio.io/testingdir ⌚ 21:54:28
$ ISTIO_NAMESPACE=istio2 make --makefile ../istio/tools/certs/Makefile linistio10-certs-istioctl 
rm: linistio10-istioctl/*: No such file or directory
fetching root ca from k8s cluster: linistio10-7caab3af9f514f028081a8180c107b69
generating linistio10/ca-key.pem
Generating RSA private key, 4096 bit long modulus
............................................................................++
........................................................................++
e is 65537 (0x10001)
generating linistio10/cluster-ca.csr
generating linistio10/ca-cert.pem
Signature ok
subject=/O=Istio/CN=Intermediate CA/L=linistio10
unable to load certificate
4732083692:error:09FFF06C:PEM routines:CRYPTO_internal:no start line:/BuildRoot/Library/Caches/com.apple.xbs/Sources/libressl/libressl-47.11.1/libressl-2.8/crypto/pem/pem_lib.c:684:Expecting: TRUSTED CERTIFICATE
make[1]: *** [linistio10/ca-cert.pem] Error 1
rm linistio10/cluster-ca.csr linistio10/intermediate.conf
make: *** [linistio10-certs-istioctl] Error 2
```

**Comment by linsun** [comment]
/hold 

@sdake have you tried this to see if it works?  see my comments above

File: tools/certs/Makefile (hunk 2 above)
Comment (costinm) [reply tools/certs/Makefile:330]: CA_ADDR is the 'central istiod address' in your case - the address of Istiod. 

It's the same address that gets configured in the pods, during injection. The steps bellow replicate what the agent is doing.

File: tools/certs/Makefile (hunk 1 above)
Comment (costinm) [reply tools/certs/Makefile:338]: You are right on the first part - it should fetch the ca_secrets from the current namespace. Also we can remove this step for both 'central istiod with proper DNS certs' as well as if we use --insecure on the grpcURL ( for testing ). 

But it should fetch it from the current namespace and using the configmap that istiod is creating.

It is WIP...

**Comment by costinm** [comment]
Ed - I don't mind if you remove the 'token' based method from Makefile. I have it in my own repo - I thought it would be useful, but I don't have time to address the comments, sorry I pushed it on you... It is unfortunate token-based doesn't have any support in istioctl - with the recent change from John ( adding JWT auth to XDS ) we'll likely no longer need certs for istioctl.

File: tools/certs/Makefile (hunk 2 above)
Comment (costinm) [reply tools/certs/Makefile:330]: Not sure what you mean about ns1/ns2 - this is the same address that a Pod would use to connect to istiod, wherever it is. 

For example I have a DNS-based cert, and my address is the URL of the gateway where Istiod is exposed ( and I don't need the cert from the config map ).

File: tools/certs/Makefile (hunk 1 above)
Comment (christian-posta) [reply tools/certs/Makefile:338]: Is the intention to always use the `default` service account for these short-lived tokens?

**Comment by esnible** [comment]
Closing in favor of https://github.com/istio/istio/pull/26513 for now.

**Comment by sdake** [comment]
@esnible - thanks. We should keep on eye on this makefile. It has almost all of the logic needed to converge all of our cert and token based logic into one. I would prefer this happens as a shell script so it Is understandable for those not strong with make.

@nmittler any thoughts - I remember we discussed this prior, but don't recall the resulting decision.

**Pull request closed without merging**